File: typescript/src/run-once.ts
/**
 * run-once — invoke an agent in batch (no HTTP, no Express lifecycle).
 *
 * The third deployment shape, complementing serving endpoints and
 * Apps hosting. Designed for Databricks Job / cron / scheduled batch
 * contexts where the agent needs to run once with a known prompt and
 * return its final text — no incoming HTTP request, no serving
 * endpoint, just compile + invoke.
 *
 * The Python equivalent (`_run_once.py`) compiles the agent via
 * `compile_to_chat_agent` and calls `predict()`. In TS, the existing
 * `runViaSDK` runner already plays the role of the compiled chat agent:
 * it takes a messages array + model and returns the final assistant
 * text. `runOnce` just builds the single-prompt input and delegates.
 *
 * `runViaSDK` shape (from `./agent/runner.js`):
 *
 *   async function runViaSDK(params: {
 *     model: string;
 *     instructions: string;
 *     messages: Array<{ role: string; content: string }>;
 *     tools: AgentTool[];
 *     subAgents?: string[];
 *     maxTurns?: number;
 *     oboHeaders: Record<string, string>;
 *     ...
 *   }): Promise<string>
 *
 * `runViaSDK` itself already extracts the last assistant message and
 * returns its content as a string — so this module's job is just
 * argument marshaling + model resolution.
 *
 * Tests inject a mock via `opts.runViaSDK` to avoid the real FMAPI call.
 */

import { z } from 'zod';
import { runViaSDK as defaultRunViaSDK } from './agent/runner.js';
import type { AgentConfig, AgentExports } from './agent/plugin.js';
import type { AgentTool } from './agent/tools.js';

// ---------------------------------------------------------------------------
// Minimal facade — narrow enough that tests can inject a mock without
// reaching for the full runner module.
// ---------------------------------------------------------------------------

export interface RunViaSDKParams {
  model: string;
  instructions: string;
  messages: Array<{ role: string; content: string }>;
  tools: AgentTool[];
  subAgents?: string[];
  maxTurns?: number;
  oboHeaders: Record<string, string>;
}

export type RunViaSDKFn = (params: RunViaSDKParams) => Promise<string>;

export type StreamViaSDKFn = (params: RunViaSDKParams) => AsyncIterable<string>;

// ---------------------------------------------------------------------------
// Options
// ---------------------------------------------------------------------------

const OptionsSchema = z.object({
  prompt: z.string().min(1, 'prompt is required and must be non-empty'),
  model: z.string().min(1).optional(),
});

export interface RunOnceOptions {
  /**
   * The agent to invoke. Accepts the `AgentExports` shape returned by
   * `createAgentPlugin(...).exports()`, a raw `AgentConfig`, or any
   * shape with `.getConfig()`/`.getTools()` accessors. The agent's
   * configured `model` is used as a fallback when `opts.model` is
   * unset.
   */
  agent: AgentExports | AgentConfig | unknown;
  /** User prompt string. Required and non-empty. */
  prompt: string;
  /** LLM endpoint. Defaults to the agent's configured `model`. */
  model?: string;
  /** Injectable runner for tests. Defaults to the real `runViaSDK`. */
  runViaSDK?: RunViaSDKFn;
}

// ---------------------------------------------------------------------------
// Agent shape narrowing
// ---------------------------------------------------------------------------

function isExports(agent: unknown): agent is AgentExports {
  return (
    typeof agent === 'object' &&
    agent !== null &&
    typeof (agent as AgentExports).getConfig === 'function'
  );
}

function isConfig(agent: unknown): agent is AgentConfig {
  return (
    typeof agent === 'object' &&
    agent !== null &&
    typeof (agent as AgentConfig).model === 'string'
  );
}

function resolveConfig(agent: unknown): AgentConfig | null {
  if (isExports(agent)) return agent.getConfig();
  if (isConfig(agent)) return agent;
  return null;
}

function resolveTools(agent: unknown, config: AgentConfig | null): AgentTool[] {
  if (isExports(agent)) return agent.getTools();
  return config?.tools ?? [];
}

// ---------------------------------------------------------------------------
// Public entry point
// ---------------------------------------------------------------------------

/**
 * Invoke an agent once with a single user prompt and return its final
 * assistant text.
 *
 * @throws ZodError when `prompt` is missing or empty.
 * @throws Error when no model can be resolved from `opts.model` or
 *   the agent's configured model.
 */
export async function runOnce(opts: RunOnceOptions): Promise<string> {
  // Validate prompt + optional model
  OptionsSchema.parse({ prompt: opts.prompt, model: opts.model });

  const config = resolveConfig(opts.agent);

  const effectiveModel = opts.model ?? config?.model;
  if (!effectiveModel) {
    throw new Error(
      'runOnce requires a model. Pass `model` explicitly or supply an ' +
        'agent with a configured `model`.',
    );
  }

  const tools = resolveTools(opts.agent, config);
  const instructions = config?.instructions ?? '';

  const runner: RunViaSDKFn = opts.runViaSDK ?? defaultRunViaSDK;

  return await runner({
    model: effectiveModel,
    instructions,
    messages: [{ role: 'user', content: opts.prompt }],
    tools,
    subAgents: config?.subAgents,
    maxTurns: config?.maxIterations,
    oboHeaders: {},
  });
}

File: python/src/apx_agent/_run_once.py
"""run_once — invoke an agent in batch (no HTTP, no FastAPI lifecycle).

The third deployment shape, complementing Model Serving (``log_agent`` +
``databricks.agents.deploy``) and Apps hosting (``create_app``). Designed
for Databricks Job / cron / scheduled batch contexts where the agent
needs to run once with a known prompt and return its final text — no
incoming HTTP request, no serving endpoint, just compile + invoke.

Today, batch users typically hand-roll the LLM loop directly against
the Model Serving endpoint (skipping the framework entirely) because no
clean batch primitive exists. That reimplements 100+ lines that the
framework already owns. ``run_once`` collapses it to a single call.

Auth: uses the framework's standard ``_make_workspace_client``
resolution. In a Databricks Job, set ``DATABRICKS_CLIENT_ID`` +
``DATABRICKS_CLIENT_SECRET`` as job env vars — the framework
auto-detects OAuth M2M. For local dev, ``DATABRICKS_TOKEN`` works.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ._agents import BaseAgent

# Top-level import so tests can patch and so import failures surface as
# friendlier errors when the eval extra (mlflow) isn't installed. The
# function itself raises a clearer error when this is None.
try:
    from ._chat_agent import compile_to_chat_agent
except ImportError:  # pragma: no cover — defensive
    compile_to_chat_agent = None  # type: ignore[assignment]

from ._inspection import _load_agent_config


def run_once(
    agent: "BaseAgent",
    prompt: str,
    *,
    model: str | None = None,
    session_id: str | None = None,
    custom_inputs: dict[str, Any] | None = None,
) -> str:
    """Invoke an agent once and return its final assistant text.

    Args:
        agent: Any ``BaseAgent`` (``LlmAgent``, ``SequentialAgent``, ...).
        prompt: User prompt string. For richer inputs (system + user, or
            a multi-turn history) call ``compile_to_chat_agent(agent)``
            directly and pass a custom ``messages`` list to ``predict``.
        model: LLM endpoint. Defaults to ``[tool.apx.agent].model`` in
            ``pyproject.toml`` (resolved via ``_load_agent_config``).
        session_id: Optional session ID — when the agent's ChatAgent has a
            session store configured, loads + persists multi-turn history.
        custom_inputs: Additional fields forwarded to ``predict()``.
            ``session_id`` is merged in if both are provided.

    Returns:
        The final assistant message text. Empty string when the agent
        produced no assistant message — rare, usually indicates the
        agent loop hit ``max_iterations`` before producing output.

    Raises:
        ValueError: when ``model`` is unset and pyproject.toml doesn't
            configure ``[tool.apx.agent].model``.

    Example::

        # Databricks Job entrypoint, scheduled via databricks.yml
        from apx_agent import run_once
        from my_agent import agent
        from my_outputs import post_to_slack

        def main() -> int:
            result = run_once(agent, "Run today's scan.")
            post_to_slack(result)
            return 0
    """
    if compile_to_chat_agent is None:
        raise ImportError(
            "run_once requires the eval extra (mlflow). "
            "Install with: pip install 'apx-agent[eval]'"
        )

    effective_model = model
    if effective_model is None:
        cfg = _load_agent_config()
        effective_model = cfg.model if cfg is not None else None
    if not effective_model:
        raise ValueError(
            "run_once requires a model. Pass model= explicitly or set "
            "[tool.apx.agent].model in pyproject.toml."
        )

    chat_agent = compile_to_chat_agent(agent, model=effective_model)

    from mlflow.types.agent import ChatAgentMessage
    messages = [ChatAgentMessage(role="user", content=prompt)]

    inputs = dict(custom_inputs or {})
    if session_id is not None:
        inputs.setdefault("session_id", session_id)

    response = chat_agent.predict(
        messages=messages,
        custom_inputs=inputs or None,
    )

    response_messages = getattr(response, "messages", []) or []
    for msg in reversed(response_messages):
        if getattr(msg, "role", None) == "assistant":
            return getattr(msg, "content", "") or ""
    return ""
